service/cachemdw/cache.go

```diff
@@ -0,0 +1,171 @@
+	decodedRequestContextKey any
+
+	*logging.ServiceLogger
+}
+
+func NewServiceCache(
+	cacheClient cache.Cache,
+	evmClient EVMClient,
+	cacheTTL time.Duration,
+	decodedRequestContextKey any,
+	logger *logging.ServiceLogger,
+) *ServiceCache {
+	return &ServiceCache{
+		cacheClient:              cacheClient,
+		evmClient:                evmClient,
+		cacheTTL:                 cacheTTL,
+		decodedRequestContextKey: decodedRequestContextKey,
+		ServiceLogger:            logger,
+	}
+}
+
+func IsCacheable(ctx context.Context, evmClient EVMClient, req *decode.EVMRPCRequestEnvelope) bool {
+	blockNumber, err := req.ExtractBlockNumberFromEVMRPCRequest(ctx, evmClient)
+	if err != nil {
+		return false
+	}
+
+	if blockNumber <= 0 {
+		return false
+	}
+
+	return true
+}
+
+func (c *ServiceCache) GetCachedQueryResponse(
+	ctx context.Context,
+	req *decode.EVMRPCRequestEnvelope,
+	host string,
+) ([]byte, error) {
+	chainID, err := c.GetAndCacheChainID(ctx, host)
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := GetQueryKey(chainID, req)
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := c.cacheClient.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return value, nil
+}
+
+func (c *ServiceCache) CacheQueryResponse(
+	ctx context.Context,
+	req *decode.EVMRPCRequestEnvelope,
+	chainID string,
+	response []byte,
+) error {
+	if !IsCacheable(ctx, c.evmClient, req) {
+		return errors.New("query isn't cacheable")
+	}
+
+	key, err := GetQueryKey(chainID, req)
+	if err != nil {
+		return err
+	}
+
+	return c.cacheClient.Set(ctx, key, response, c.cacheTTL)
+}
+
+func (c *ServiceCache) ValidateAndCacheQueryResponse(
+	ctx context.Context,
+	req *decode.EVMRPCRequestEnvelope,
+	host string,
+	response []byte,
+) error {
+	// TODO(yevhenii): add validation
+
+	chainID, err := c.GetAndCacheChainID(ctx, host)
+	if err != nil {
+		return err
+	}
+
+	if err := c.CacheQueryResponse(
+		ctx,
+		req,
+		chainID,
+		response,
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ServiceCache) GetCachedChainID(
+	ctx context.Context,
+	host string,
+) (string, error) {
+	key := GetChainKey(host)
+
+	chainID, err := c.cacheClient.Get(ctx, key)
+	if err != nil {
+		return "", err
+	}
+
+	return string(chainID), nil
+}
+
+func (c *ServiceCache) GetAndCacheChainID(
+	ctx context.Context,
+	host string,
+) (string, error) {
+	chainID, err := c.GetCachedChainID(ctx, host)
+	if err != nil && err != cache.ErrNotFound {
+		return "", err
+	}
+	if err == cache.ErrNotFound {
+		rawChainID, err := c.evmClient.ChainID(ctx)
+		if err != nil {
+			return "", err
+		}
+
+		err = c.CacheChainID(ctx, host, rawChainID.String())
+		if err != nil {
+			return "", err
+		}
+
+		chainID = rawChainID.String()
+	}
+
+	return chainID, nil
+}
+
+func (c *ServiceCache) CacheChainID(
+	ctx context.Context,
+	host string,
+	chainID string,
+) error {
+	key := GetChainKey(host)
+
+	return c.cacheClient.Set(ctx, key, []byte(chainID), c.cacheTTL)
+}
```

Comment:
I understand the code, but not why its here -> please update the architecture docs with what this service is responsible for / how it works conceptually

Reply:
basically I don't think we should have a caching service running in the background, it's too much work / complexity for very little payoff. If we cache chainid when a user requests it using the standard caching flow, that should be more than good enough/

Reply:
ChainID caching is removed as discussed